Log origin of client commands.

hjoliver opened the pull request:
Something else BoM SMS operators like to see: logging of which user issued a command. This is more important for SMS where one server program runs all suites ... but cylc can still have several users sharing a passphrase to the same suite. This change just passes user@host from the client command to the server command (and info) interface, and logs it:

```
2015-03-27T10:31:38 INFO - Command from oliverh@wrh-1.niwa.co.nz: trigger task(foo,1,False)
2015-03-27T10:31:39 INFO - Command succeeded: trigger task(foo,1,False)
2015-03-27T10:31:40 INFO - [foo.1] -initiate job-submit
```

[update: Close #1421]

hjoliver commented:
My test battery passes.

hjoliver commented:
Note that is a simple interim measure in lieu of proper multi-user authentication, which we plan to do  after the communications refactor.

This branch should provide a convenient log of who did what for suites in which multiple users have the suite passphrase, but it would be possible for a trusted user (i.e. one who has the passphrase) to fake the username that gets logged with his/her commands.

hjoliver commented:
Rebased onto master.

hjoliver commented:
Rebased again.

hjoliver commented:
@matthewrmshin - please review.

hjoliver commented:
Note that in order to centralize the new command logging I had to add an argument to the command queue ```put()``` method, which affects a lot of commands - having to modify these gave me the opportunity to factor out a bunch of repeated exception handling code as well.

matthewrmshin commented:
Have you lost some logic for `bin/cylc-broadcast` after merging #1388?

hjoliver commented:
I don't think so - why do you ask?  The new broadcast tests pass, and the file is almost identical on this branch and current master:
```
% git diff master -- bin/cylc-broadcast | cat
diff --git a/bin/cylc-broadcast b/bin/cylc-broadcast
index ccfde52..0629e9c 100755
--- a/bin/cylc-broadcast
+++ b/bin/cylc-broadcast
@@ -32,7 +32,7 @@ from cylc import cylc_pyro_client
 from cylc.command_prep import prep_pyro
 from cylc.print_tree import print_tree
 from cylc.task_id import TaskID
-from parsec.validate import validate, ValidationError
+from parsec.validate import validate
 from cylc.cfgspec.suite import SPEC
 import cylc.flags
```

hjoliver commented:
(my full test battery passes on this branch)

matthewrmshin commented:
(Just a bit confused, because every other commands are using `from cylc.cylc_pyro_client import client`.)

matthewrmshin commented:
Test battery is indeed OK.

matthewrmshin commented:
Broadcast does not appear to have the client's user@host logged.

matthewrmshin commented:
Others are working OK.

hjoliver commented:
A-ha, I see what you mean!  Broadcast is in a sense not a "command" in that it does not go through the suite command interface - there is a separate Pyro-connected interface just for broadcasts (the reason for this escapes me at the moment - hopefully there was a good reason!).  But you're right that user@host should also be logged for any broadcasts, so I'll implement that...

hjoliver commented:
Rebased onto master.

hjoliver commented:
@matthewrmshin - I've finished this one off.  My test battery passes.

matthewrmshin commented:
This is now looking really good. Some minor points:
* Do we need an automated test or 2 for the new functionality?
* I thought I would do a `git grep '\.get_proxy('` to see what's still connecting to the suite daemon without logging `user@host`. The call from `cylc.task_message` is probably fine? `cylc dump`, `cylc monitor` and `cylc gui` are all getting `state_summary`, although `cylc gui` also connects for `suite-info` and `log`. I wonder if it is worth logging `user@host` for some of these activities as well? For `cylc monitor` and `cylc gui`, perhaps we can log the initial connect? (And perhaps use `atexit` to log a disconnect as well?)

hjoliver commented:
I've added a test, will look at your other suggestions too...

matthewrmshin commented:
New test passes in my environment.

hjoliver commented:
Close #1421.

hjoliver commented:
(note this branch is currently in a broken transitional state - I'm part way through adding proper logging of all client requests)

hjoliver commented:
Branch squashed and rebased onto master.  My test battery passes.

The new client connection logging now reports the client name, user-at-host, and a UUID to distinguish between instances of the same client (e.g. different gcylc instances).  Multi-connection clients (GUI and monitor) now report sign-in and sign out. The GUI reports each individual connection to the suite command interface, but not suite-info and suite-state interfaces unless ``` --debug```.  The GUI and monitor put their client UUID in the title bar, and CLI clients can optionally print their UUID to stderr.

```
% cylc cat-log foo | grep Client
2015-05-06T17:15:34+12 INFO - Client sign-in (cylc-gui oliverh@wrh-1.niwa.co.nz e50b4d61-c3ac-4cba-bfe3-b272565b21c7)
2015-05-06T17:15:52+12 INFO - Client command hold_suite (cylc-gui oliverh@wrh-1.niwa.co.nz e50b4d61-c3ac-4cba-bfe3-b272565b21c7)
2015-05-06T17:16:02+12 INFO - Client sign-in (cylc-monitor oliverh@wrh-1.niwa.co.nz 2eefcb09-155f-44f8-9ef4-9bfffcc451fe)
2015-05-06T17:16:09+12 INFO - Client command release_suite (cylc-gui oliverh@wrh-1.niwa.co.nz e50b4d61-c3ac-4cba-bfe3-b272565b21c7)
2015-05-06T17:16:15+12 INFO - Client command trigger_task (cylc-gui oliverh@wrh-1.niwa.co.nz e50b4d61-c3ac-4cba-bfe3-b272565b21c7)
2015-05-06T17:16:35+12 INFO - Client sign-out (cylc-monitor oliverh@wrh-1.niwa.co.nz 2eefcb09-155f-44f8-9ef4-9bfffcc451fe)
2015-05-06T17:16:39+12 INFO - Client sign-out (cylc-gui oliverh@wrh-1.niwa.co.nz e50b4d61-c3ac-4cba-bfe3-b272565b21c7)
2015-05-06T17:16:48+12 INFO - Client command set_stop_cleanly (cylc-stop oliverh@wrh-1.niwa.co.nz b852aa45-3a1c-4632-ac1a-0e1d20f65f7e)
```

This touched a lot of the Pyro client/server code and associated "suite interfaces", much of which is pretty old and crusty, and I found it hard to follow, so a bit of refactoring was needed:

 * The Pyro code and all suite client/server interfaces are now held in a "network" package directory.  For each interface, the client and server classes - which need to have matching methods - are in the same file.
 * The gcylc updater connect/disconnect/reconnect code was pretty confused, so I refactored that to make it much more straightforward. With ```gcylc --debug``` the GUI prints connection info to stderr so you can see exactly what is happening in this regard.

This isolates and simplifies the Pyro code quite a bit, so it should make the upcoming comms refactor easier.

Compatibility with older suite daemons checked.

matthewrmshin commented:
@benfitzpatrick As discussed, I am happy with this change, with tests pass in my environment. Given the scale of this change, however, I would like another pair of eyes to sanity check.

benfitzpatrick commented on bin/cylc-depend:
```diff
@@ -61,18 +61,10 @@ else:
+pclient = SuiteCommandClient(
+    suite, pphrase, options.owner, options.host, options.pyro_timeout,
+    options.port)
+pclient.put_command('add_prerequisite', task_id, msg)
```
This change and the others are nice. Hopefully this is close to the kind of interface we'll need for the generic communications refactor.

benfitzpatrick commented on lib/cylc/gui/app_gcylc.py:
```diff
@@ -548,11 +550,7 @@ def __init__(self, suite, db, owner, host, port, pyro_timeout,
-            title += " -r (restricted display)"
-        self.window.set_title(title)
-
+        self.window.set_title('gcylc - %s' % self.cfg.my_uuid)
```
I really don't like the UUID in the gcylc window title - can it be moved to the 'About' dialog or something?

hjoliver replied:
Ok, I'll put it in 'About' and under the 'View' menu.

benfitzpatrick commented on lib/cylc/network/client_reporter.py:
```diff
@@ -0,0 +1,98 @@
+        self.clients = {}
+
+    def report(self, command, uuid, info, multi):
+        now = datetime.datetime.now()
```
Better to use `utcnow` to avoid jumps

hjoliver replied:
Yep.

benfitzpatrick commented on lib/cylc/network/suite_state.py:
```diff
@@ -107,55 +113,65 @@ def update( self, tasks, tasks_rh, min_point, max_point, max_point_rh, paused,
         self.task_summary = task_summary
         self.global_summary = global_summary
         self.family_summary = family_summary
         task_states = {}
```
`self.first_update_completed = True` here, instead?

hjoliver replied:
Yep, done.

benfitzpatrick commented on lib/cylc/scheduler.py:
```diff
@@ -159,65 +165,30 @@ def __init__( self, is_restart=False ):
-                }
+        self.info_commands = {}
+        for attr in dir(self):
+            if attr.startswith('info_'):
```
I would be tempted to put in a `callable(` test here

hjoliver replied:
OK.

benfitzpatrick commented on lib/cylc/scheduler.py:
```diff
@@ -251,15 +222,13 @@ def configure( self ):
 
-        # LOAD TASK POOL ACCORDING TO STARTUP METHOD
+        self.suite_state = StateSummaryServer(
+            self.config, self.run_mode, str(self.initial_point))
```
I had thought that the change from `str(self.pool.get_min_point())` to `str(self.initial_point)` might cause problems during a restart, but it turns out only 5.4.x gcylcs actually use the resulting `start time` property. They won't work for a whole host of other reasons with cylc 6 suites anyway. Can we remove this third argument and the `start time` global summary key?

hjoliver replied:
Yeah, it's not being used anywhere ... removed.

benfitzpatrick commented:
The `SuiteStillInitialisingError` traceback from `cylc gui` is a bit scary - can we get rid of the traceback and just print a warning message?

benfitzpatrick commented:
My graph view keeps spamming `graph raw` exceptions for one suite.

benfitzpatrick commented:
For a suite that looks like this:
```
[scheduling]
    [[dependencies]]
        graph = "foo => bar"
```
then a simple `cylc run` and `cylc gui` gives the `graph raw` complaint.

benfitzpatrick commented on lib/cylc/gui/updater.py:
```diff
@@ -135,174 +148,198 @@ def __init__(self, cfg, info_bar, restricted_display):
+
+    def update(self):
+        if cylc.flags.debug:
+            print >> sys.stderr, "UPDATE", ctime().split()[3],
```
Better to use full date-time and timezone?

hjoliver replied:
I think minimal timing info is fine here - this debug trace is just to allow developers to see what happens in what order when the gui connects, disconnects, and reconnects to a suite daemon.

hjoliver commented:
> My graph view keeps spamming graph raw exceptions for one suite.

Fixed. One of my final changes on this branch was to remove the unnecessary indirection in client commands (each command method had a name that the client used, now we just use the method name) ... I had missed two client calls in the graph updater.

hjoliver commented:
> The SuiteStillInitialisingError traceback from cylc gui is a bit scary - can we get rid of the traceback and just print a warning message?

Oops, I had only tested this with the monitor.  I've added a popup progress bar for this - see what you think of it.  Also, popup warnings for no connection, and connection lost.

hjoliver commented:
@benfitzpatrick - all feedback points addressed, plus:
  * gcylc - new popup dialogs and a progress bar for connection status
  * back-compat restored (I had broken it again with the aforementioned command indirection fix).

hjoliver commented:
Test battery still passes.

benfitzpatrick commented:
The popup progress bar is very nice - it just needs an icon in the dialog window and needs the parent window to be set (otherwise it pops up in a random bit of the screen). If you set `set_type_hint` on the window to `gtk.gdk.WINDOW_TYPE_HINT_SPLASHSCREEN` then it wouldn't have any icon, title, or button decoration, which might be appropriate.

The warning dialogs are useful when you're doing something unusual like opening a gcylc for a stopped suite.

However, I get two warning dialogs when a suite stops under normal circumstances - connection lost, and reconnection failed. Could we maybe replace these with some info in the status bar, or drop them altogether in this situation?

benfitzpatrick commented:
I get this backwards compatibility error for a ~6.4.0 daemon:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python2.6/threading.py", line 532, in __bootstrap_inner
    self.run()
  File "/my/path/to/cylc/lib/cylc/gui/updater.py", line 470, in run
    and self.update()):
  File "/my/path/to/cylc/lib/cylc/gui/updater.py", line 373, in update
    self.retrieve_state_summaries()
  File "/my/path/to/cylc/lib/cylc/gui/updater.py", line 275, in retrieve_state_summaries
    if glbl['stopping']:
KeyError: 'stopping'
```

benfitzpatrick commented:
(I only get that error sometimes, actually)

hjoliver commented:
I decided a popup progress bar was a bit silly; it's now embedded in the info bar, and it is also active when the suite is "stopping".  I think that's useful because "stopping" is a condition that can take some time but cannot be backed out of even though the suite appears to be still running normally (apart from the word "stopping" in the status widget, which is comparatively simple).  See what you think...

I got rid of the warning dialog for "connection lost", which was unnecessary since the gui immediately tries to reconnect and spawns a "port file not found - suite not running?" warning.  However, I've kept the latter warning.  So now you get one of those at gui start-up if the suite is not running, and one if the suite stops while the gui is up.  I think this is a good thing for two reasons: (1) it's the same message you get from the CLI if the suite isn't running; and (2) shows the user exactly why the gui thinks it isn't running - port file missing - which demystifies the workings of the gui somewhat.   Are you persuaded :grimacing:

hjoliver commented:
> I get this backwards compatibility error for a ~6.4.0 daemon

This is because at <= 6.4.0 the state summary object isn't Pyro-connected until all tasks are loaded (on this branch I connect it early, but raise the new exception until initialisation is done).  At 6.4.0 this kills the monitor, but the gui reports the suite is stopped"(which isn't correct, but better than aborting).  I should be able to handle this, but I've run out of time for the moment...

hjoliver commented:
[update: done]

benfitzpatrick commented:
I like the progress bar for the initialising bit (although it is slightly too tall if you use medium icons on our system - see the screenshot). The stopping progress bar might be annoying if your suite takes a long time to stop (e.g. waiting for a running task).

![screenshot](https://cloud.githubusercontent.com/assets/1138333/7532772/dec3554c-f562-11e4-9138-3db4254f160d.png)

I'm not sure about an extra warning dialog for a normal stop - given that the info bar says 'stopped with X', it looks a bit confusing to say 'suite not running?'.

benfitzpatrick commented:
Maybe this information is better in the bottom left of the info bar?

hjoliver commented:
>  it is slightly too tall if you use medium icons ...

The progress bar is now added without expansion inside a VBox, so it doesn't expand to the high of large icons.  However, it'll always be higher than small icons due to the text inside it.

> The stopping progress bar might be annoying if your suite takes a long time to stop ...

Progress bars are meant to be annoying (or attention-grabbing at least) in order to alert the user to some important transitional state that prevents them from seeing or doing what they can expect to see or do under normal circumstances.  I think stopping fits that bill - it looks like business as usual in the GUI but you can no longer do certain things because the non-reversible shutdown process has started.  However, as advertised by tool tip, the easily annoyed can now click on the progress bar to make it disappear! 

> Maybe this information is better in the bottom left of the info bar?

Yeah, I agree.  I had put it in the middle to avoid making the permanent status info widget jump around, but actually we don't need to see both at the same time.  So now the progress bar temporarily replaces the status info widget on the far left.   This works well IMO.

hjoliver commented:
> I'm not sure about an extra warning dialog for a normal stop - given that the info bar says 'stopped with X', it looks a bit confusing to say 'suite not running?'.

I've now prevented the warning dialog for an expected/normal stop.  It still pops up in the event of unexpected suite death, and at GUI start-up when the suite isn't running, but with a bit more information in it (alluding to the fact that "stopped with X" really means "cannot communicate with the suite so I'm assuming it has stopped" - which isn't necessarily true).

hjoliver commented:
The progress bar is now active during reload too.  And some new logging:
```
% cylc log foo | grep Reload
2015-05-13T12:29:26+12 WARNING - Reload will complete once current active tasks have finished.
2015-05-13T12:29:46+12 INFO - Reload completed.
```
I've had occasional complaints that it's hard to know when a reload is finished, in a big suite.

hjoliver commented:
(BTW I moved the progress bar GUI code from the updater into the main app module, which is where it should be.)

hjoliver commented:
(Rebased last few commits to remove I file that I accidentally added from another dev branch).

hjoliver commented:
(test battery still passes)

hjoliver commented:
@benfitzpatrick - I think I've addressed all your concerns about the final changes on this branch, namely: warning dialog on normal shutdown - gone; progress bar position - moved to the left; progress bar may be annoying - click to remove it.  Progress bar activity on reload is new.  I'd like to  get this in soon if possible as it's big enough that it conflicts with every new merge :grimacing:
